### ISSUE TITLE
Maybe it's reasonable to recommend wrapping a `callbackFlow` initialization into `try`-`finally`?

### ISSUE THREAD

**Issue by andrew-k-21-12** [Kotlin/kotlinx.coroutines#4206]
This is just a suggestion, most likely it's quite controversial, but anyway. 

**What do we have now?**

Now the docs for `callbackFlow` provide the following code example:

```kotlin
fun flowFrom(api: CallbackBasedApi): Flow<T> = callbackFlow {
    val callback = object : Callback { // Implementation of some callback interface
        override fun onNextValue(value: T) {
            // To avoid blocking you can configure channel capacity using
            // either buffer(Channel.CONFLATED) or buffer(Channel.UNLIMITED) to avoid overfill
            trySendBlocking(value)
                .onFailure { throwable ->
                    // Downstream has been cancelled or failed, can log here
                }
        }
        override fun onApiError(cause: Throwable) {
            cancel(CancellationException("API Error", cause))
        }
        override fun onCompleted() = channel.close()
    }
    api.register(callback)
    /*
     * Suspends until either 'onCompleted'/'onApiError' from the callback is invoked
     * or flow collector is cancelled (e.g. by 'take(1)' or because a collector's coroutine was cancelled).
     * In both cases, callback will be properly unregistered.
     */
    awaitClose { api.unregister(callback) }
}
```

**What should be instead?**

Maybe it's reasonable to change it into the following instead (just schematically, to describe the point):

```kotlin
fun flowFrom(api: CallbackBasedApi): Flow<T> = callbackFlow {
    val callback = { ... }
    try {
        api.register(callback)
    } finally {
        awaitClose { api.unregister(callback) }
    }
}
```

**Why?**

Any unexpected exception at `api.register(callback)` (or similar initialization) makes the `callback` leaked. Such leak can be quite difficult to detect because when catching exceptions from `Flow`s (i.e. `flowFrom(api).catch { ... }`) an exception from `api.register(callback)` can be just ignored and suppressed.

**Why not?**

More boilerplate code. The situation described above can be super rare to happen, so such `try`-`finally` blocks will be just an overkill for nothing.

**Comment by dkhalanskyjb** [comment]
If `register` throws, there are two possibilities, depending on how `register` is implemented:

1. `callback` was registered anyway.
2. `callback` was not registered.

The way the sample is now, case 2) is properly covered: `register` threw an exception, and the coroutine closes, as if `onApiError` was called. Case 1) is not properly covered, yes: the code will work correctly, but a callback will be leaked.

The proposed change to the sample means that case 1) is somewhat properly covered: if the callback was registered despite the exception and then is properly completed, we'll wait for the callback. 2), however, is no longer covered: this coroutine will just hang indefinitely in `awaitClose`, as there is no one to close the channel.

I think 2) is a much more sensible assumption: typically, when APIs throw exceptions, they try to avoid mutating any shared state. Why throw an exception if the mutation that was requested actually succeeded?

**Comment by andrew-k-21-12** [comment]
> 2), however, is no longer covered: this coroutine will just hang indefinitely in awaitClose, as there is no one to close the channel.

Got the point. Yes, the code I provided in the **What should be instead?** section is incomplete and should be:

```kotlin
fun flowFrom(api: CallbackBasedApi): Flow<T> = callbackFlow {
    val callback = { ... }
    try {
        api.register(callback)
    } catch (throwable: Throwable) {
        close(throwable)
    } finally {
        awaitClose { api.unregister(callback) }
    }
}
```

> typically, when APIs throw exceptions, they try to avoid mutating any shared state. Why throw an exception if the mutation that was requested actually succeeded?

Yes, agree. My points here are:
1. Just a human mistake. There can be an unexpected exception inside `api.register(callback)` after the `callback` has been registered. Such situations shouldn't happen in the ideal case but if they will, there will be not very noticeable callback leak.
2. Other quite rare system errors. For example, `OutOfMemoryError`. In this case this error will be propagated to the `callbackFlow { ...}`'s collection. If this collection will have a `catch` block, this `OutOfMemoryError` will be just silenced with a one more callback leak.

**Comment by dkhalanskyjb** [comment]
W.r.t. `OutOfMemoryError`: if your program is at a point where there isn't enough memory for it to continue running, one or two leaked callbacks are the least of your concerns, but luckily, the program is likely to crash soon in this scenario anyway.

As for human mistakes: sure, those happen, but when `register` starts throwing exceptions, I'd assume that it does something else incorrectly as well. Maybe `unregister` also throws. Maybe the callbacks arrive to the wrong consumers. Who knows, really, what happens when the system is in a clearly incorrect state? Trying to salvage the situation and keep the program running correctly in case there are mistakes in the code is difficult, and resource leaks, again, shouldn't be the top priority.

Is there any specific API where you noticed this callback leaks happening, or are your concerns purely theory-based? I'm all for improving the documentation samples, but I'm wary about including a suggestion that our users overcomplicate their code for no actual benefit.

**Comment by andrew-k-21-12** [comment]
> if your program is at a point where there isn't enough memory for it to continue running, one or two leaked callbacks are the least of your concerns, but luckily, the program is likely to crash soon in this scenario anyway.

True. But if the program is configured to restore its situation with memory in cases of `OutOfMemoryError`s, there will be leaked callbacks coming from such `callbackFlow`s.

> Trying to salvage the situation and keep the program running correctly in case there are mistakes in the code is difficult, and resource leaks, again, shouldn't be the top priority.

The program will be running in any case if there is a `catch` block for collection of such `callbackFlow` (if it doesn't re-throw the exception, of course), it will just have one more memory leak. One mistake (even much bigger one) should not be a justification for another mistake.

I understand all points against the suggestion. I kept them in mind - this is why I wrote "most likely it's quite controversial" in the starting post. 

Nevertheless, I can't provide any real world examples of APIs having such callback leaks happening. So it can seem rational to wait for other reports (having more practical details) before introduction of a complication for the documentation.
